src/async_impl/response.rs

```diff
@@ -240,6 +240,56 @@ impl Response {
         serde_json::from_slice(&full).map_err(crate::error::decode)
     }
 
+    /// Try to deserialize a single chunk of the request body as JSON.
+    ///
+    /// When the response body has been exhausted, this will return `Ok(None)`.
+    ///
+    /// # Optional
+    ///
+    /// This requires the optional `json` feature enabled.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # extern crate reqwest;
+    /// # extern crate serde;
+    /// #
+    /// # use reqwest::Error;
+    /// # use serde::Deserialize;
+    /// #
+    /// // This `derive` requires the `serde` dependency.
+    /// #[derive(Deserialize)]
+    /// struct Ip {
+    ///     origin: String,
+    /// }
+    ///
+    /// # async fn run() -> Result<(), Error> {
+    /// let mut res = reqwest::get("http://httpbin.org/ip").await?;
+    ///
+    /// while let Some(chunk) = res.json_chunk::<Ip>().await? {
+    ///     println!("Chunk: {:?}", chunk);
+    /// }
+    /// # Ok(())
+    /// # }
+    /// ```
+    ///
+    /// # Errors
+    ///
+    /// This method fails whenever the response chunk is not in JSON format
+    /// or it cannot be properly deserialized to target type `T`. For more
+    /// details please see [`serde_json::from_reader`].
+    ///
+    /// [`serde_json::from_reader`]: https://docs.serde.rs/serde_json/fn.from_reader.html
+    #[cfg(feature = "json")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "json")))]
+    pub async fn json_chunk<T: DeserializeOwned>(&mut self) -> crate::Result<Option<T>> {
+        if let Some(full) = self.chunk().await? {
+            Ok(Some(serde_json::from_slice(&full).map_err(crate::error::decode)?))
```

Comment:
So, this is assuming that every "chunk", which is basically a `read()` call on the socket, is a full JSON message? It's a fragile assumption, since the data could be combined into a single read, or it could be too large and be broken up into multiple reads...

Reply:
> So, this is assuming that every "chunk" [...] is a full JSON message?

Yes

> It's a fragile assumption, since the data could be combined into a single read, or it could be too large and be broken up into multiple reads...

The assumption is that `chunk()` corresponds to one chunk from [`Transfer-Encoding: chunked`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding). Not that it corresponds to whatever the OS decides `read()` is.

If this assumption is incorrect, then `reqwest` probably needs to be extended to support `Transfer-Encoding: chunked`, because otherwise I can't consume token streams from [`text-generation-inference`](https://github.com/huggingface/text-generation-inference).

Reply:
> reqwest (and hyper) does support chunked transfer-encoding. But it doesn't buffer up every "chunk" that way. The decoder in hyper will do 1 OS read, and then pass on either up to the chunked delimiter, or the full thing if the delimiter is not yet reached.

How do you suggest I tell when a delimiter is reached, then? Is there a method for this?

Reply:
Chunked delimiters have no semantic significance, so they can also be changed as they go through gateways/proxies. I'm not familiar with the project you linked. Most often, JSON streaming is done by delimiting JSON objects with newlines. Then what you do, is read and buffer until you get a newline, and then you can decode the object.

Reply:
> Chunked delimiters have no semantic significance, so they can also be changed as they go through gateways/proxies.

This is a server-side issue, isn't it? If the server hosts an endpoint that doesn't use the encoding properly, that's a bug on their end.

If `reqwest` needs to be resilient to this sort of issue, I would still want to support endpoints that actually work properly - so we'd end up having two families of functions, one that uses chunks and one that uses newlines.

In that case, I'll still need a way to tell the end of a chunk apart from an arbitrary read boundary.

Is there any way to test for this?

Reply:
No, there isn't a way to test for it. The properties of the transfer are something that hyper handles, they are not considered part of the content (vs `Content-Encoding`).

It's sort of similar to how calling `read()` on an OS TcpStream won't tell you which bytes were in each segment, since the OS will combine segments into a single buffer if it receives multiple in-between calls to `read()`.

Reply:
Here's my plan - let's remove the new `json_chunk` methods from this PR and narrow the scope to just adding that blocking `chunk` method - and then I'll work on some Server-Sent Events support in a separate PR.

It looks like reqwest's `Decoder` doesn't support it - we can read chunks until `\n\n`, but we can't just give portions of the chunk back to the `Decoder` if the server happens to send multiple events in a single chunk; if chunking is an implementation detail and we have to be agnostic, then we have to support that scenario.

Maybe we can provide an iterator over each event in an SSE stream? Then the iterator can keep track of whether the last event ended in the middle of some chunk.